src/wp-includes/block-bindings/class-wp-block-bindings.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Block Bindings API: WP_Block_Bindings class.
+ *
+ * Support for overriding content in blocks by connecting them to different sources.
+ *
+ * @package WordPress
+ * @subpackage Block Bindings
+ */
+
+/**
+ * Core class used to define supported blocks, register sources, and populate HTML with content from those sources.
+ *
+ *  @since 6.5.0
+ */
+class WP_Block_Bindings {
+
+	/**
+	 * Holds the registered block bindings sources, keyed by source identifier.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @var array
+	 */
+	private $sources = array();
+
+	/**
+	 * Function to register a new block binding source.
+	 *
+	 * Sources are used to override block's original attributes with a value
+	 * coming from the source. Once a source is registered, it can be used by a
+	 * block by setting its `metadata.bindings` attribute to a value that refers
+	 * to the source.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string   $source_name   The name of the source.
+	 * @param array    $source_args   The array of arguments that are used to register a source. The array has two elements:
+	 *                                1. string   $label        The label of the source.
+	 *                                2. callback $apply        A callback
+	 *                                executed when the source is processed during
+	 *                                block rendering. The callback should have the
+	 *                                following signature:
+	 *
+	 *                                  `function (object $source_attrs, object $block_instance, string $attribute_name): string`
+	 *                                          - @param object $source_attrs: Object containing source ID used to look up the override value, i.e. {"value": "{ID}"}.
+	 *                                          - @param object $block_instance: The block instance.
+	 *                                          - @param string $attribute_name: The name of an attribute used to retrieve an override value from the block context.
+	 *                                 The callback should return a string that will be used to override the block's original value.
+	 *
+	 * @return void
+	 */
+	public function register_source( string $source_name, array $source_args ) {
+		$this->sources[ $source_name ] = $source_args;
+	}
+
+	/**
+	 * Depending on the block attributes, replace the HTML based on the value returned by the source.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $block_content Block content.
+	 * @param string $block_name The name of the block to process.
+	 * @param string $block_attr The attribute of the block we want to process.
+	 * @param string $source_value The value used to replace the HTML.
+	 */
+	public function replace_html( string $block_content, string $block_name, string $block_attr, string $source_value ) {
```

Comment:
I've removed the public `replace_html()` method as it shouldn't be exposed as public API because there's no need for it. This is following @gziolo 's comments in https://github.com/WordPress/wordpress-develop/pull/5888#pullrequestreview-1838271936

Reply:
So, since the method is private we won't test it directly but we can test it indirectly by calling `$block_instance->render()` for example.